src/pages/posts/[slug].astro

```diff
@@ -0,0 +1,18 @@
+---
+import { getCollection } from 'astro:content';
+
+export async function getStaticPaths() {
+  const postsEntries = await getCollection('posts');
+  return postsEntries.map(entry => ({
+    params: { slug: entry.slug }, props: { entry },
+  }));
+}
```

Comment:
📝 Currently SSR not support custom 404 page using `src/pages/404.astro`. So I use static output mode.